src/common/address.py

```diff
@@ -36,10 +36,24 @@ def from_absolute(cls, address: str):
         return cls(path, address, protocol)
 
     @classmethod
-    def from_relative(cls, address: str, document_id: str | None, data_source: str):
+    def from_relative(
+        cls, address: str, document_id: str | None, data_source: str, relative_path: list[str] | None = None
+    ):
         if "://" in address:
             # Already contains protocol and data source
             return cls.from_absolute(address)
+        if address.startswith("~"):
+            if not relative_path:
+                raise ApplicationException(
+                    "Missing the relative path and therefore it is not possible to resolve the ~ inside the address."
+                )
+            go_up = address.count("~")
+            path = relative_path.copy()
```

Comment:
Might be good enough for now. We there should at least be a check on unsupported use of `"~"`.
relative references like ".data.~.data" will not work now. Which probably is fine.

Can just check that if there is any number of "~.", they must all be at the start of the reference